File: sorting/MergeSort/java/MergeSort.java
/* All copyrights reserved by Venkys.io.
 Visit https://venkys.io for more information */

// Java program for implementing Merge Sort
// This program demonstrates divide and conquer sorting algorithm

// Space complexity: O(n)
// Time complexity:
//|Operation    | Best Case | Average Case | Worst Case |
//|-------------|-----------|--------------|------------|
//|Merge Sort   | O(nlogn)  | O(nlogn)     | O(nlogn)   |

import java.util.Arrays;

public class MergeSort {
    // Function to merge two sorted sub-arrays
    static void merge(int[] arr, int left, int mid, int right) {
        int n1 = mid - left + 1;    // Size of left subarray
        int n2 = right - mid;       // Size of right subarray

        // Create temporary arrays
        int[] L = new int[n1];      // Left subarray
        int[] R = new int[n2];      // Right subarray

        // Copy data to temporary arrays
        for (int i = 0; i < n1; i++)
            L[i] = arr[left + i];
        for (int j = 0; j < n2; j++)
            R[j] = arr[mid + 1 + j];

        // Merge temporary arrays back into arr[left..right]
        int i = 0, j = 0, k = left;
        while (i < n1 && j < n2) {
            if (L[i] <= R[j]) {
                arr[k] = L[i];      // Copy from left array
                i++;
            } else {
                arr[k] = R[j];      // Copy from right array
                j++;
            }
            k++;
        }

        // Copy remaining elements of L[]
        while (i < n1) {
            arr[k] = L[i];
            i++;
            k++;
        }

        // Copy remaining elements of R[]
        while (j < n2) {
            arr[k] = R[j];
            j++;
            k++;
        }
    }

    // Function to implement Merge Sort
    static void mergeSort(int[] arr, int left, int right) {
        if (left < right) {
            int mid = left + (right - left) / 2;  // Find middle point

            mergeSort(arr, left, mid);            // Sort first half
            mergeSort(arr, mid + 1, right);       // Sort second half
            merge(arr, left, mid, right);         // Merge sorted halves
        }
    }

    // Function to print array elements
    static void printArray(int[] arr) {
        System.out.println(Arrays.toString(arr));
    }

    // Main method to demonstrate merge sort
    public static void main(String[] args) {
        int[] arr = {12, 11, 13, 5, 6, 7};
        System.out.println("Original array:");
        printArray(arr);

        mergeSort(arr, 0, arr.length - 1);

        System.out.println("Sorted array:");
        printArray(arr);
    }
}

//Sample Input: [12, 11, 13, 5, 6, 7]
//Sample Output: [5, 6, 7, 11, 12, 13]

/* Join our communities*/
// Follow us on LinkedIn - https://www.linkedin.com/company/venkysio
// Follow us on Facebook - https://www.facebook.com/venkysio
// Follow us on Instagram - https://www.instagram.com/venkys.io
// Follow us on Twitter - https://twitter.com/iovenkys
// Follow us on Youtube - https://www.youtube.com/@CoreCodersNetwork

File: sorting/MergeSort/python/MergeSort.py
# All copyrights reserved by Venkys.io.
# Visit https://venkys.io for more information

# Python program for implementing Merge Sort
# This program demonstrates divide and conquer sorting algorithm

# Space complexity: O(n)
# Time complexity:
#|Operation    | Best Case | Average Case | Worst Case |
#|-------------|-----------|--------------|------------|
#|Merge Sort   | O(nlogn)  | O(nlogn)     | O(nlogn)   |

# Function to merge two sorted sub-arrays
def merge(arr, left, mid, right):
    n1 = mid - left + 1    # Size of left subarray
    n2 = right - mid       # Size of right subarray

    # Create temporary arrays
    L = arr[left:mid + 1]  # Left subarray
    R = arr[mid + 1:right + 1]  # Right subarray

    # Merge temporary arrays back into arr[left..right]
    i = j = 0             # Initial indexes of left and right subarrays
    k = left             # Initial index of merged subarray

    while i < n1 and j < n2:
        if L[i] <= R[j]:
            arr[k] = L[i]  # Copy from left array
            i += 1
        else:
            arr[k] = R[j]  # Copy from right array
            j += 1
        k += 1

    # Copy remaining elements of L[]
    while i < n1:
        arr[k] = L[i]
        i += 1
        k += 1

    # Copy remaining elements of R[]
    while j < n2:
        arr[k] = R[j]
        j += 1
        k += 1

# Function to implement Merge Sort
def merge_sort(arr, left, right):
    if left < right:
        mid = (left + right) // 2  # Find middle point

        merge_sort(arr, left, mid)        # Sort first half
        merge_sort(arr, mid + 1, right)   # Sort second half
        merge(arr, left, mid, right)      # Merge sorted halves

# Function to print array elements
def print_array(arr):
    print(" ".join(map(str, arr)))

# Main function to demonstrate merge sort
if __name__ == "__main__":
    arr = [12, 11, 13, 5, 6, 7]
    print("Original array:")
    print_array(arr)

    merge_sort(arr, 0, len(arr) - 1)

    print("Sorted array:")
    print_array(arr)

#Sample Input: [12, 11, 13, 5, 6, 7]
#Sample Output: [5, 6, 7, 11, 12, 13]

# Join our communities
# Follow us on LinkedIn - https://www.linkedin.com/company/venkysio
# Follow us on Facebook - https://www.facebook.com/venkysio
# Follow us on Instagram - https://www.instagram.com/venkys.io
# Follow us on Twitter - https://twitter.com/iovenkys
# Follow us on Youtube - https://www.youtube.com/@CoreCodersNetwork
